File: year2015/day01/python/solution.py
def main():
    print(part_1())
    print(part_2())


def part_1():
    sum = 0
    for char in input:
        sum += 1 if char == '(' else -1
    return sum


def part_2():
    return next((index + 1 for index, element in enumerate(input) if find_index_condition(element, index, input)), None)


def setup(path):
    global input
    
    with open(path, 'r') as file:
        input = [*file.read()]


def find_index_condition(element, index, array):
    return sum(1 if c == '(' else -1 for c in array[:index + 1]) == -1


if __name__ == '__main__':
    setup('../puzzle.txt')
    main()

File: year2015/day01/java/Solution.java
import java.nio.file.*;
import java.util.*;

public class Solution {
    private static String[] CHARS;

    public static void main(String[] args) { 
        try {
            setup(Files.readString(Paths.get("../puzzle.txt")));
            System.out.println(part1());
            System.out.println(part2());
        } catch(Exception e) {
            e.printStackTrace();
        }
    }
    
	public static Integer part1() {
        int sum = 0;
        for (String ch : CHARS) {
            sum += ch.equals("(") ? 1 : -1;
        }
        return sum;
    }

    public static Integer part2() {
        for (int i = 0, sum = 0; i < CHARS.length; i++) {
            sum += CHARS[i].equals("(") ? 1 : -1;
            if (sum == -1) {
                return i + 1;
            }
        }
        return -1;
    }

    public static void setup(String input) {
        CHARS = input.split("");
    }
}